### ISSUE TITLE
Use BOVC, BNVC to improve add with overflow checks on r6.

### ISSUE THREAD

**Issue by paul99** [v8mips/v8mips#201]
In certain cases, add & sub of Smi's or ints must be checked for overflow, and special handling or deopts be performed. For example, see `LCodeGen::DoAddI()`.

Traditional MIPS has trapping add/sub instructions, but we decided to avoid use of traps and signal handlers in v8. (There is complexity, possibly security/sandboxing issues, and degenerate cases where the trap-to-kernel overhead could tank performance.)

The alternative was to compute overflow independently, as is done with the macro instructions: `MacroAssembler::AdduAndCheckForOverflow()`, `MacroAssembler::SubuAndCheckForOverflow`, `MacroAssembler::BranchOnOverflow() and related/variant instructions. (There are multiple versions, to optimize the sequence of 3 instructions and minimize register usage.) 

While profiling Octane benchmark it was found that this overhead of overflow computation took significant performance hit in some tests. It was found experimentally that the checking of overflow on Adds turned out to be significantly more common than checking Subs in the workloads that were tested.

Instructions were added to r6 to minimize the overhead to check. (BOVC & BNVC).

In both old and new variations, we must execute the addu() instruction anyway to actually do the computation, and store the result. With BOVC/BNVC, we can reduce the check-overflow-and-branch sequence from 5 instructions to 1. (these instructions do have forbidden slots so they do require guards.)

Note that these instructions take the 2 source arguments in registers, so variants of AdduAndCheckForOverflow() that use constant-operand will have to load those values to a temp register.

It is very likely that the current 'semantic' of AdduAndCheckForOverflow() and BranchOnOverflow() will not properly model both the oldand new methods, and a new approach will be needed. It would be preferable (but maybe not mandatory) to continue to hide these variations in the macro-assembler.

**Comment by kilvadyb** [comment]
Optimal solutions in pre-R6 and R6 are quite different for me, I do not see a nice way to hide release differences in MacroAssembler efficiently. Inefficient way would be to set `overflow_dst` of `MacroAssembler::AdduAndCheckForOverflow()` with using `bovc` but in that case the R6 implementation almost as big and slow as in pre-R6. Other idea is to use DeoptimizIfPre() and DeoptimizeIfPost() functions (but that would hurt the common interface IMO). Please compare the current and R6 optimized DoAddI functions:

in pre-R6 (current impl.):

```
void LCodeGen::DoAddI(LAddI* instr) {
  LOperand* left = instr->left();
  LOperand* right = instr->right();
  LOperand* result = instr->result();
  bool can_overflow = instr->hydrogen()->CheckFlag(HValue::kCanOverflow);

  if (!can_overflow) {
    if (right->IsStackSlot()) {
      Register right_reg = EmitLoadRegister(right, at);
      __ Addu(ToRegister(result), ToRegister(left), Operand(right_reg));
    } else {
      DCHECK(right->IsRegister() || right->IsConstantOperand());
      __ Addu(ToRegister(result), ToRegister(left), ToOperand(right));
    }
  } else {  // can_overflow.
    Register overflow = scratch0();
    Register scratch = scratch1();
    if (right->IsStackSlot()) {
      Register right_reg = EmitLoadRegister(right, scratch);
      __ AdduAndCheckForOverflow(ToRegister(result),
                                 ToRegister(left),
                                 right_reg,
                                 overflow);  // Reg at also used as scratch.
    } else {
      DCHECK(right->IsRegister() || right->IsConstantOperand());
      __ AdduAndCheckForOverflow(ToRegister(result), ToRegister(left),
                                 ToOperand(right), overflow, scratch);
    }
    DeoptimizeIf(lt, instr, Deoptimizer::kOverflow, overflow,
                 Operand(zero_reg));
  }
}
```

And in R6:

```
void LCodeGen::DoAddI(LAddI* instr) {
  LOperand* left = instr->left();
  LOperand* right = instr->right();
  LOperand* result = instr->result();
  bool can_overflow = instr->hydrogen()->CheckFlag(HValue::kCanOverflow);

  if (can_overflow) {
    // ...
    // Setting up right_reg.
    // ....
    DeoptimizeIf(bovc, instr, Deoptimizer::kOverflow, left, Operand(right_reg));
  } else {
    if (right->IsStackSlot()) {
      Register right_reg = EmitLoadRegister(right, at);
      __ Addu(ToRegister(result), ToRegister(left), Operand(right_reg));
    } else {
      DCHECK(right->IsRegister() || right->IsConstantOperand());
      __ Addu(ToRegister(result), ToRegister(left), ToOperand(right));
    }
  }
}
```

Currently my only idea is to use DeoptimizeIfPre() and DeoptimizeIfPost() and ...Pre() is a nop in pre-R6 while in R6 ...Post() is a nop and `AdduAndCheckForOverflow()` just calls `Addu()`:

```
void LCodeGen::DoAddI(LAddI* instr) {
  LOperand* left = instr->left();
  LOperand* right = instr->right();
  LOperand* result = instr->result();
  bool can_overflow = instr->hydrogen()->CheckFlag(HValue::kCanOverflow);

  if (!can_overflow) {
    if (right->IsStackSlot()) {
      Register right_reg = EmitLoadRegister(right, at);
      __ Addu(ToRegister(result), ToRegister(left), Operand(right_reg));
    } else {
      DCHECK(right->IsRegister() || right->IsConstantOperand());
      __ Addu(ToRegister(result), ToRegister(left), ToOperand(right));
    }
  } else {  // can_overflow.
    Register overflow = scratch0();
    Register scratch = scratch1();
    DeoptimizeIfPre(bovc, instr, Deoptimizer::kOverflow, overflow,
                 Operand(zero_reg));
    if (right->IsStackSlot()) {
      Register right_reg = EmitLoadRegister(right, scratch);
      __ AdduAndCheckForOverflow(ToRegister(result),
                                 ToRegister(left),
                                 right_reg,
                                 overflow);  // Reg at also used as scratch.
    } else {
      DCHECK(right->IsRegister() || right->IsConstantOperand());
      __ AdduAndCheckForOverflow(ToRegister(result), ToRegister(left),
                                 ToOperand(right), overflow, scratch);
    }
    DeoptimizeIfPost(lt, instr, Deoptimizer::kOverflow, overflow,
                 Operand(zero_reg));
  }
}
```

**Comment by kilvadyb** [comment]
Ákos had a nice idea to use true and fals Label\* parameters for SubuAndCheckForOverflow() and AdduAndCheckForOverflow() functions. I checked the lithium and turbofan codes and it seems to be possible to use *AndCheckForOverflow() functions that way. And DeoptimizeIf() will be called with `al` condition in lithium code. I am working on a proposal/test implementation.

**Comment by kilvadyb** [comment]
Seems if I don't want to make heavy changes in turbofan code then two different function must be used for lithium and turbofan to achieve some almost optimal code for pre-r6 and r6.

**Comment by lialan** [comment]
A side discussion here:

while reading the code AdduAndCheckForOverflow on MIPS32 I can see that in the case of adding an immediate value, we are using bitwise operations to detect overflow. I have done a very similar optimization in MagicCode regarding it:

Assume incoming immediate is a 32bit unsigned value, and then if we regard it as a signed value, we will have the following observation (extracted from MagicCode source code comments I made a while ago, and op1 is from the "left" register, and op3 is from the "dst" register in V8 AdduAndCheckForOverflow function context. And actually, it works for SubuAndCheckForOverflow as well):

```
using (op1, op2, op3) to represent the MSB of (operand_reg, operand_imm32, result_reg) respecitvely.
when opnd_imm32 >=0 then the calculation of
(op1 ^ op3) & (op2 ^ op3) could be simplified as (op1 ^ op3) & op3

when opnd_imm32 < 0 then it becomes: (op1 ^ op3) & ~op3.
since we only care for the MSB, use a truth table could give us:
op1, op3  => (op1 ^ op3) & ~op3
0 , 0   => 0
0 , 1   => 0
1 , 0   => 1
1 , 1   => 0
```

It can shrink the comparison instructions from 3 to 2 in both cases to calculate the overflow bit, and better, it can avoid restoring immediate value in the "right" operand to a temporary register to use op2 can be used to do an XOR operation, because op2 is omitted in the computation.

I haven't seen the whole picture of how this macro fits into the whole codegen component yet, but I think it might work in a similar case in v8. I will try to make a proof of concept for this.

**Comment by paul99** [comment]
Very interesting, Alan, I need to think about that a little bit. PTAL at https://codereview.chromium.org/1434263003/ and feel free to review or contribute ideas.

**Comment by kilvadyb** [comment]
Alan: sounds good, although if new branch op (with nop) must be used for immediate signs then woman  have the same instruction amount at the end. I suggest to start a new CL for this proposal work.

**Comment by kilvadyb** [comment]
I was thinking about using `bovc, bnvc` for `SubuAndCheckForOverflow`: If we can negate one of the operands (`subu(scratch, zero_reg, left);`) then we cold use `bovc(scratch, right, overflow_label);`. But if the value in left register is MIN_INT then negating it also causes an overflow. And if we have to check register values then we are at the same size like the pre-r6 implementation.

**Comment by kilvadyb** [comment]
I saw that the MIPS64 port also contains the new functions like `DaddBranchNoOvf` and the old functions (DadduAndCheckForOverflow) as well. I think the old function interface should be dropped and the existing code should be cleaned up.

**Comment by kilvadyb** [comment]
32-bit version is landed in 21b331e3ff9f4da302d31627db7ca265adf1a767 commit, 64 bit is continued in #227
